cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,114 +285,690 @@ func copyFromGlobHelper(src, dst, globEx string) error {
++---+---+           +-------+----+                +-------+----+                        |               |
+|       |           | mount      |       success  | mount data |                        |               |
+| done  +<----------+ save       |      +---------+            +<-----------------------+               |
+|       |           |            |      |         |            |                                        |
++--+----+           +----+-------+      |         +----------+-+                                        |
+   ^                     ^              |                    |                                          |
+   |                     | success      v                    |                                          |
+   |                     |     +--------+----+   fail        |fail                                      |
+   |                     |     | unlock save +--------+      |                                          |
+   |                     +-----+ w/ run key  |        v      v                                          |
+   |                     ^     +-------------+   +----+------+-----+                                    |
+   |                     |                       | unlock save     |                                    |
+   |                     |                       | w/ fallback key +------------------------------------+
+   |                     +-----------------------+                 |
+   |                             success         +-------+---------+
+   |                                                     |
+   |                                                     |
+   |                                                     |
+   +-----------------------------------------------------+
+						fail
+
+*/
+
+type stateMachine struct {
+	// the current state is the one that is about to be executed
+	current stateFunc
+
+	// the disk we have all our partitions on
+	disk disks.Disk
+
+	// TODO: roll these next two into degradedState instead ?
+
+	// the device for ubuntu-data that is to be mounted - either an unencrypted
+	// volume, or a decrypted mapper volume if ubuntu-data is really encrypted
+	// on the physical disk
+	dataDevice string
+	// the device for ubuntu-save that is to be mounted
+	saveDevice string
+
+	isEncryptedDev bool
+
+	// state for tracking what happens as we progress through degraded mode of
+	// recovery
+	degradedState recoverDegradedState
+}
+
+func (m *stateMachine) diskOpts() *disks.Options {
+	if m.isEncryptedDev {
+		return &disks.Options{
+			IsDecryptedDevice: true,
+		}
+	}
+	return nil
+}
+
+func (m *stateMachine) verifyMountPoint(dir, name string) error {
+	matches, err := m.disk.MountPointIsFromDisk(dir, m.diskOpts())
 	if err != nil {
 		return err
 	}
+	if !matches {
+		return fmt.Errorf("cannot validate boot: %s mountpoint is expected to be from disk %s but is not", name, m.disk.Dev())
```

Comment:
tweaked to `fmt.Errorf("cannot validate mount: %s mountpoint %s is expected to be from disk %s but is not", name, dir, m.disk.Dev())`

Reply:
this is a bit confusing because is the target of the mountpoint that should come from that disk. The issue is probably that the helper itself has a bit of an ambiguous name.  Maybe MountIsFromDisk would be clearer?

anyway here, s/mountpoint/mountpoint target/ perhaps?

Reply:
renamed to stateMachine.mountIsFromDisk